server/src/main/java/org/elasticsearch/script/field/WriteField.java

```diff
@@ -282,8 +394,80 @@ public Object get(Object defaultValue) {
         if (leaf == null) {
             return defaultValue;
         }
+        return getFromContainer(leaf, defaultValue);
+    }
 
-        return container.getOrDefault(leaf, defaultValue);
+    public Object get() {
+        Object value = get(MISSING);
+        throwExceptionIfPresent();
+        if (value == MISSING) {
+            throw new IllegalArgumentException("cannot resolve path [" + path + "]");
+        }
+        return value;
+    }
+
+    public void throwExceptionIfPresent() {
+        Optional.ofNullable(error).ifPresent(e -> { throw e.get(); });
+    }
+
+    @SuppressWarnings("unchecked")
+    private Object getFromContainer(String leaf, Object defaultValue) {
+        if (container instanceof Map<?, ?>) {
+            Map<Object, Object> mapContainer = (Map<Object, Object>) container;
+            if (mapContainer.containsKey(leaf) == false) {
+                setError(() -> new IllegalArgumentException("field [" + leaf + "] not present as part of path [" + path + "]"));
+            }
+            return mapContainer.getOrDefault(leaf, defaultValue);
+        } else if (container instanceof List<?> listContainer) {
+            int index = parseIndex(leaf);
+            if (index >= 0 && index < listContainer.size()) {
+                return listContainer.get(index);
+            } else {
+                setError(
+                    () -> new IllegalArgumentException(
+                        "["
+                            + index
+                            + "] is out of bounds for array with length ["
+                            + listContainer.size()
+                            + "] as part of path ["
+                            + path
+                            + "]"
+                    )
+                );
+                return defaultValue;
+            }
+        } else {
+            if (container == null) {
+                setError(() -> new IllegalArgumentException("cannot resolve [" + leaf + "] from null as part of path [" + path + "]"));
+            } else {
+                setError(
+                    () -> new IllegalArgumentException(
+                        "cannot resolve [" + leaf + "] from object of type [" + typeName(container) + "] as part of path [" + path + "]"
+                    )
+                );
+            }
+            return defaultValue;
+        }
+    }
+
+    private void setError(Supplier<IllegalArgumentException> errorSupplier) {
+        if (error == null) {
+            error = errorSupplier;
+        }
+    }
```

Comment:
This bridges the semantical differences between the fields API that never really throws an exception when accessing a field that doesn't exist with the IngestDocument API that is very strict when referencing a field that doesn't exist. The accessor methods can register an error and the caller can decide whether they want a strict or a lenient behavior.